### ISSUE TITLE
Any advice how to load/evaluate JavaScript code after the include?

### ISSUE THREAD

**Issue by fbrandel** [gustafnk/h-include#40]
Is there any advice / best practice how to evaluate the JavaScript code (inline or external) which is part of the response? It is not an h-include issue but I think it would be helpful to point it out somewhere.

**Comment by gustafnk** [comment]
Hi Florian,

There's a tiny comment on that topic deep buried in the [Why not HTML Imports? section](https://github.com/gustafnk/h-include#why-not-html-imports). I should definitely clarify that.

In general, I would look at how [jquery-pjax](https://github.com/defunkt/jquery-pjax) loads scripts. But, I would recommend against it, since I think script loading from HTML responses is quite a can-of-worms.

Look for example here: https://github.com/defunkt/jquery-pjax

Though, if you only have one script to include, things probably gets significantly simpler :)

I'd love to get more details of what you want to achieve. Do you have any source to share?

**Comment by fbrandel** [comment]
The use case here is basically including a (micro)service which brings its own CSS and JS in its content. The path to these resources might not be known by the "outer" document.  Doing this via Edge Side Includes is not a real problem – things start to get tricky when the content is loaded async on the client because the scripts are not evaluated.

Two possible solutions come into mind:
- Hook into one of the h-include hooks, search for script-tags and include/evaluate these in the "outside" document.
- Include the needed script resources in the "outside" document, so the scripts are already executed safely. But this would need some knowledge/coupling between the components which should be avoided.

**Comment by gustafnk** [comment]
I'd opt for trying the second solution + sending Custom Elements in the HTML response. Then, the only knowledge shared is what kind of custom elements that exist in the system as a whole.

Is it the same team that is building the consuming service and the microservice? If so, I think you're in a good position.

What kind of script is it? In many cases, you need a contract between the services anyway (i.e. a `<div id="consumer-insert-content-here />` or `<div class="class-that-the-consumer-knows-about" />`)

**Comment by fbrandel** [comment]
In a best case scenario the consumer would only know about the URL of the microservice, which it includes via `<h-include>`. The microservice can deliver its `<fancy-thing>` element or whatever including the corresponding JS functionality.

If the consumer service would know about `<fancy-thing>` it would have to change if `<another-fancy-thing>` element appears. That is manageable but not ideal ;)

Maybe the contract between consumer and microservice could be that the microservice always has a script.js file which can be included by the consumer, regardless if it is needed or not.
Or the microservice tells the consumer what resources to include (maybe as part of a HTTP header field)

**Comment by gustafnk** [comment]
Hmmm, but if you have such a `script.js` contract, you will not be able to cache bust. So the cache might be valid for a long time and "unknown" custom elements may appear in the response.

However, you might use a script loader for this. It's creates a small performance overhead though.

Check out the section on script loading in my [Microservice Websites article](https://gustafnk.github.io/microservice-websites/#local-stylesheets-and-scripts). Could that approach with for example little-loader help in your scenario?

**Comment by fbrandel** [comment]
Your right.This convention could bring up even more problems, so we skipped that for now. We now fetch the needed scripts from the microservice DOM and load them just before the DOM gets inserted into the consuming service. Looks good (and simple) for now. Unfortunately we did not extend `h-include`, so I can currently not show any example which could help other people...

Feel free to close this issue, the discussion helped a lot! 
And your article was very helpful, too 👍
